File: app/src/gui/helpers/DeletionWorker.py
import os
import shutil 

from PySide6.QtWidgets import QMessageBox
from PySide6.QtCore import QThread, Signal, QWaitCondition, QMutex # Added QWaitCondition, QMutex
from ...utils.definitions import SUPPORTED_IMG_FORMATS


class DeletionWorker(QThread):
    progress = Signal(int, int)  # (deleted, total)
    finished = Signal(int, str)  # (count, message)
    error = Signal(str)
    
    # NEW SIGNAL: Requests confirmation from the main thread
    confirm_signal = Signal(str, int) # (message, total_items)

    def __init__(self, config):
        super().__init__()
        self.config = config
        self.confirmation_response = False # Stores Yes/No reply
        self.wait_condition = QWaitCondition() # Used to pause execution
        self.mutex = QMutex() # Used to protect shared variables

    # NEW: Slot to receive the user's decision from the main thread
    def set_confirmation_response(self, response: bool):
        self.mutex.lock()
        self.confirmation_response = response
        self.mutex.unlock()
        self.wait_condition.wakeOne() # Resume the worker thread

    def run(self):
        try:
            target_path = self.config["target_path"]
            mode = self.config.get("mode", "files") 
            require_confirm = self.config["require_confirm"]

            if not target_path or not os.path.exists(target_path):
                self.error.emit("Target path does not exist.")
                return

            # --- DIRECTORY DELETION MODE ---
            if mode == 'directory':
                if not os.path.isdir(target_path):
                    self.error.emit(f"Error: Target path is not a directory: {target_path}")
                    return
                
                if require_confirm:
                    msg = f"Permanently delete the directory and all its contents: \n\n{target_path}\n\nThis cannot be undone!"
                    
                    self.mutex.lock()
                    self.confirm_signal.emit(msg, 1) # Emit signal to main thread
                    self.wait_condition.wait(self.mutex) # PAUSE worker until main thread responds
                    self.mutex.unlock()

                    if not self.confirmation_response:
                        self.finished.emit(0, "Directory deletion cancelled by user.")
                        return

                self.progress.emit(0, 1) 
                try:
                    shutil.rmtree(target_path)
                    self.finished.emit(1, f"Successfully deleted directory and its contents: {target_path}")
                except Exception as e:
                    self.error.emit(f"Failed to delete directory {target_path}: {e}")
                
                return 

            # --- FILE DELETION MODE (Existing Logic) ---
            
            extensions = self.config["target_extensions"] or SUPPORTED_IMG_FORMATS
            
            exts = {f".{ext.lstrip('.').lower()}" for ext in extensions}
            files_to_delete = []

            if os.path.isfile(target_path):
                if any(target_path.lower().endswith(ext) for ext in exts):
                    files_to_delete.append(target_path)
            else:
                for root, _, files in os.walk(target_path):
                    for file in files:
                        if any(file.lower().endswith(ext) for ext in exts):
                            files_to_delete.append(os.path.join(root, file))

            total = len(files_to_delete)
            if total == 0:
                self.finished.emit(0, "No files found matching the selected extensions.")
                return

            if require_confirm:
                msg = f"Permanently delete {total} file(s) matching extensions?\n\nThis cannot be undone!"
                
                self.mutex.lock()
                self.confirm_signal.emit(msg, total) # Emit signal to main thread
                self.wait_condition.wait(self.mutex) # PAUSE worker until main thread responds
                self.mutex.unlock()

                if not self.confirmation_response:
                    self.finished.emit(0, "Deletion cancelled by user.")
                    return

            deleted = 0
            for file_path in files_to_delete:
                try:
                    os.remove(file_path)
                    deleted += 1
                    self.progress.emit(deleted, total)
                except Exception as e:
                    print(f"Failed to delete {file_path}: {e}")

            self.finished.emit(deleted, f"Deleted {deleted} file(s).")
            
        except Exception as e:
            self.error.emit(str(e))

File: app/src/gui/helpers/DeletionWorker.cpp
#include "DeletionWorker.h"
#include "core/FSETool.h" // Assumed C++ equivalent
#include "core/FileDeleter.h" // Assumed C++ equivalent
#include <QFileInfo>
#include <QDir>
#include <QSet>

// Define supported formats (assumed from Python)
const QStringList DeletionWorker::SUPPORTED_IMG_FORMATS = {
    "png", "jpg", "jpeg", "bmp", "webp", "gif"
};

DeletionWorker::DeletionWorker(const QVariantMap& config, QObject* parent)
    : QThread(parent), m_config(config), m_confirmationResponse(false)
{
}

void DeletionWorker::setConfirmationResponse(bool response)
{
    QMutexLocker locker(&m_mutex);
    m_confirmationResponse = response;
    m_waitCondition.wakeOne();
}

void DeletionWorker::run()
{
    try {
        QString targetPath = m_config["target_path"].toString();
        QString mode = m_config.value("mode", "files").toString();
        bool requireConfirm = m_config["require_confirm"].toBool();

        if (targetPath.isEmpty() || !QFileInfo::exists(targetPath)) {
            emit error("Target path does not exist.");
            return;
        }

        // --- DIRECTORY DELETION MODE ---
        if (mode == "directory") {
            if (!QFileInfo(targetPath).isDir()) {
                emit error(QString("Error: Target path is not a directory: %1").arg(targetPath));
                return;
            }
            
            if (requireConfirm) {
                QString msg = QString("Permanently delete the directory and all its contents: \n\n%1\n\nThis cannot be undone!").arg(targetPath);
                
                m_mutex.lock();
                emit confirmSignal(msg, 1);
                m_waitCondition.wait(&m_mutex);
                m_mutex.unlock();

                if (!m_confirmationResponse) {
                    emit finished(0, "Directory deletion cancelled by user.");
                    return;
                }
            }

            emit progress(0, 1);
            
            if (FileDeleter::deletePath(targetPath)) {
                emit finished(1, QString("Successfully deleted directory and its contents: %1").arg(targetPath));
            } else {
                emit error(QString("Failed to delete directory %1").arg(targetPath));
            }
            return; 
        }

        // --- FILE DELETION MODE ---
        QStringList extensions = m_config["target_extensions"].toStringList();
        if (extensions.isEmpty()) {
            extensions = SUPPORTED_IMG_FORMATS;
        }
        
        QSet<QString> exts;
        for (const QString& ext : extensions) {
            exts.insert(QString(".%1").arg(ext.remove(0, 1).toLower()));
        }
        
        QSet<QString> filesToDeleteSet;
        QFileInfo targetInfo(targetPath);
        
        if (targetInfo.isFile()) {
            QString suffix = targetInfo.suffix().toLower();
            if (!suffix.startsWith('.')) suffix.prepend('.');
            if (exts.contains(suffix)) {
                filesToDeleteSet.insert(targetPath);
            }
        } else {
            // Recursively search the directory
            for (const QString& ext : extensions) {
                QStringList files = FSETool::getFilesByExtension(targetPath, ext, true);
                filesToDeleteSet.unite(QSet<QString>(files.begin(), files.end()));
            }
        }

        QStringList filesToDelete = QList<QString>::fromSet(filesToDeleteSet);
        filesToDelete.sort();

        int total = filesToDelete.length();
        if (total == 0) {
            emit finished(0, "No files found matching the selected extensions.");
            return;
        }

        if (requireConfirm) {
            QString msg = QString("Permanently delete %1 file(s) matching extensions?\n\nThis cannot be undone!").arg(total);
            
            m_mutex.lock();
            emit confirmSignal(msg, total);
            m_waitCondition.wait(&m_mutex);
            m_mutex.unlock();

            if (!m_confirmationResponse) {
                emit finished(0, "Deletion cancelled by user.");
                return;
            }
        }

        int deleted = 0;
        for (const QString& filePath : filesToDelete) {
            if (FileDeleter::deletePath(filePath)) {
                deleted += 1;
            }
            emit progress(deleted, total);
        }
        
        emit finished(deleted, QString("Deleted %1 file(s).").arg(deleted));
        
    } catch (const std::exception& e) {
        emit error(e.what());
    }
}